codebasin/preprocessor.py

```diff
@@ -544,15 +530,15 @@ class Visit(Enum):
     NEXT_SIBLING = 1
 
 
+@dataclass(eq=False)
 class Node:
     """
     Base class for all other Node types.
     Contains a single parent, and an ordered list of children.
     """
 
-    def __init__(self):
-        self.children = []
-        self.parent = None
+    children: list[Self] = field(default_factory=list, init=False)
+    parent: Self | None = field(default=None, init=False)
```

Comment:
Do you need `field(...)` if the default is None?

i.e. `parent: Self | None = None` should be suffice I think?

Reply:
The reason to use `field(...)` here (and https://github.com/intel/code-base-investigator/pull/144#discussion_r1940281432) is the `init=False`.

Without `init=False`, these end up exposed as parameters to the `__init__()` function.  The way that `@dataclass` inheritance works, you end up with a `parent` parameter in the `__init__()` function of every derived class, and that prevents derived classes from defining any non-default parameters (because it's unclear how you would order them).

It's definitely a bit weird, but I convinced myself that expressing things this way highlights that all `Node` objects have these special members, but it's impossible to initialize them. The only way to modify these fields is to access them directly, after a `Node` has been created.